### ISSUE TITLE
advertise address not respected for RPC

### ISSUE THREAD

**Issue by watsonian** [hashicorp/nomad#11075]
### Nomad version
```
Nomad v1.1.3+ent (47b3ee18939d473f03c26699fdde65a6dae985a1)
```

### Operating system and Environment details
```
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=21.04
DISTRIB_CODENAME=hirsute
DISTRIB_DESCRIPTION="Ubuntu 21.04"
```

Small 2 core, 1GB vagrant VM.

### Issue

The [`advertise` stanza](https://www.nomadproject.io/docs/configuration#advertise) allows you to specify which IP is advertised for other clients to use when on a machine with multiple interfaces and thus multiple possibilities. This allows you to specify a specific address for `http`, `rpc`, and `serf` which then dictates what's shown when Nomad starts:

```
nomad-server-1 nomad[1562]: Advertise Addrs: HTTP: 172.28.128.73:4646; RPC: 172.28.128.73:4647; Serf: 172.28.128.74:4648
```

Unfortunately, Nomad doesn't seem to respect the `RPC` address that's set if it differs from the `Serf` address. When you run `nomad operator raft list-peers` in the above environment, you get the following:

```
$ nomad operator raft list-peers
Node                   ID                  Address             State     Voter  RaftProtocol
nomad-server-2.global  172.28.128.75:4647  172.28.128.75:4647  leader    true   2
nomad-server-1.global  172.28.128.74:4647  172.28.128.74:4647  follower  true   2
nomad-server-3.global  172.28.128.78:4647  172.28.128.78:4647  follower  true   2
```

As you can see there, `server-1` is showing `172.28.128.74` as the RPC address even though the startup logs show that it should be `172.28.128.73`. This is reflected in client consul discovery:

```
2021-08-23T15:47:07.452Z [INFO]  client.consul: discovered following servers: servers=[172.28.128.75:4647, 172.28.128.74:4647, 172.28.128.78:4647]
```

As near as I can tell, Nomad is using the address used by `Serf` regardless of what you set the advertise address to even though the startup log line shows the correct IP and the correct IP is registered in Consul:

![image](https://user-images.githubusercontent.com/244/130510012-ab63a960-e5e1-4140-b016-154f91655c5e.png)

### Reproduction steps

Start a Vagrant VM with a definition similar to this:

```ruby
config.vm.define "nomad-server" do |server|
  server.vm.box = "bento/ubuntu-21.04"
  server.vm.hostname = "nomad-server"
  server.vm.network "private_network", type: "dhcp"
  server.vm.network "private_network", type: "dhcp"

  server.vm.provider "virtualbox" do |vb|
    vb.memory = "1024"
  end
end
```

Start the nomad server with the following settings:

```
bind_addr = "0.0.0.0"

advertise {
    http = "{{ GetInterfaceIP \"eth1\" }}"
    rpc  = "{{ GetInterfaceIP \"eth1\" }}"
    serf = "{{ GetInterfaceIP \"eth2\" }}"
}
```

#### Expected Result

Nomad should use the IP from `eth1` as the advertised RPC address and that should be reflected in `nomad operator raft list-peers` output and client consul discovery.

#### Actual Result

Nomad uses the IP from `eth2`.

**Comment by schmichael** [comment]
Thanks for the report @watsonian!

If you or somebody else has a use case this is blocking I'd love to hear it. While it's a bug I'd love to have fixed, it's more likely to get done sooner if there's not a workaround.

**Comment by pinkahd** [comment]
Is there an update to this issue or a workaround? I've encountered the same issue while we're deploying Nomad at MUBI.

We were using the fallowing Nomad config:

```
bind_addr  = "0.0.0.0"

advertise {
  http = "{{ GetInterfaceIP \"tailscale0\" }}"
  rpc  = "{{ GetInterfaceIP \"tailscale0\" }}"
  serf = "{{ GetInterfaceIP \"tailscale0\" }}"
}

consul {
  address               = "127.0.0.1:8500"
...
...
```

Once we've joined more Consul nodes the cluster got unstable and the consul checks started to toggle between a failed and active state.

Our aim was to have Nomad running on the host but to only advertise itself under the `tailscale` network (VPN).
